lib/model/folder_sendrecv.go

```diff
@@ -567,17 +567,39 @@ func (f *sendReceiveFolder) handleDir(file protocol.FileInfo, dbUpdateChan chan<
 
 	info, err := f.fs.Lstat(file.Name)
 	switch {
-	// !!!
-	// This is wrong: It deletes the file on disk regardless of
-	// what it is (e.g. got updated -> conflict)
-	// !!!
-	// There is already something under that name, but it's a file/link.
-	// Most likely a file/link is getting replaced with a directory.
-	// Remove the file/link and fall through to directory creation.
-	case err == nil && (!info.IsDir() || info.IsSymlink()):
-		err = osutil.InWritableDir(f.fs.Remove, f.fs, file.Name)
+	// There is already something under that name, we need to handle that.
+	// Unless it already is a directory, as we only track permissions,
+	// that don't result in a conflict.
+	case err == nil && !info.IsDir():
+		// Check that it is what we have in the database.
+		curFile, hasCurFile := f.model.CurrentFolderFile(f.folderID, file.Name)
+		if changed, err := f.itemChanged(info, curFile, hasCurFile, scanChan); err != nil {
+			f.newPullError(file.Name, err)
+			return
+		} else if changed {
+			l.Debugln("item changed on disk compared to db; not replacing with dir:", file.Name)
+			scanChan <- curFile.Name
+			f.newPullError(file.Name, errModified)
+			return
+		}
+
+		// Remove it to replace with the dir.
+		if !curFile.IsSymlink() && f.inConflict(curFile.Version, file.Version) {
```

Comment:
Shouldn't this check info.IsSymlink()?

Reply:
I guess it has not changed, so we can trust the db?

Reply:
Also the not a dir not a symlimk is spread over two ifs making it had to follow the logic.

Reply:
The two checks aren't really related: If `IsDir`, the item on disk doesn't need to be checked/conflicted/deleted/archived, it just gets updated. If `!IsDir`, we need to check `!IsSymlink` due to conflicts (only then). Would a (redundant) `curFile.IsDirectory()` make it more readable?

Reply:
I think adding a second directory check before moving for conflict would make it clearer. Also, I am starting to wonder how this works, the local conflict is always the loosing conflict which is not what I would have expected.

Reply:
That's admittedly a bit magical: In the db layer files are ordered by the usual conflict mechanism, and are thus only needed locally if our side of the conflict is loosing. That's why we don't need to check which item wins in the puller - if it wasn't the local one it wouldn't come up in the puller.